### DISCUSSION TITLE
Memory leak detected in BtyeBuf.release()

### DISCUSSION THREAD

**Discussion by dlvenable** [line/armeria#5367] [General]
We use Armeria and recently saw some error logs like the following:

```
2023-12-27T23:16:48.364 [armeria-common-worker-epoll-3-2] ERROR       io.netty.util.ResourceLeakDetector - LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
```

Is this a bug in Armeria? Could this be caused by a programming error in our code? And if so, how do we best debug this type of situation?

We are currently using Armeria 1.25.2.

Here is our internal issue: https://github.com/opensearch-project/data-prepper/issues/3912

**Comment by ikhoon** [comment]
A known leak has been fixed in https://github.com/line/armeria/pull/5227 with Armeria 1.26.0
It is difficult to know the cause of the leak with the given information. Could you try to raise the [leak detection level](https://netty.io/wiki/reference-counted-objects.html#leak-detection-levels) to find the leak locations?

**Reply by dlvenable** [reply]
@ikhoon , Thank you for the response. We will update the Armeria version to see if that helps. And we'll raise the leak detection level to get some more information.

At a high-level, can our code be responsible for this? I don't see our code needing to use the `ByteBuf`, but can we cause this by handling the request incorrectly?

**Reply by ikhoon** [reply]
If you didn't use `ByteBuf` directly by subscribing to `HttpRequest` with `WITH_POOLED_OBJECTS`, it would be a bug in the Armeria internal.
https://github.com/line/armeria/blob/74eb55f77507a9f6f06d37fd88c2098b69a78185/core/src/main/java/com/linecorp/armeria/common/stream/SubscriptionOption.java#L43-L49
